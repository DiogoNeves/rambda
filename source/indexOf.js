import { equals } from './equals'
import {_isIterable} from './_internals/isIterable'
import { _isArray } from './_internals/_isArray'

export function _indexOf(valueToFind, list) {
  if (!_isArray(list)){
    throw new Error(`Cannot read property 'indexOf' of ${list}`)
  }
  if (!_isIterable(valueToFind)) return list.indexOf(valueToFind)

  let index = -1
  list.forEach((x, i) => {
    if (index > -1) return
    if (equals(x, valueToFind)) index = i
  })

  return index
}


export function indexOf(valueToFind, list){
  if (arguments.length === 1){
    return _list => _indexOf(valueToFind, _list)
  }

  return _indexOf(valueToFind, list)
}
