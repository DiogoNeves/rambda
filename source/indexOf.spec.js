import { indexOf } from './indexOf'

test('without list of objects - no R.equals', () => {
  expect(indexOf(3, [ 1, 2, 3, 4 ])).toEqual(2)

  expect(indexOf(10)([ 1, 2, 3, 4 ])).toEqual(-1)
})

test('list of objects uses R.equals', () => {
  const listOfObjects = [
    {a:1},
    {b:2},
    {c:3}
  ]
  expect(indexOf({c:4}, listOfObjects)).toBe(-1)
  expect(indexOf({c:3}, listOfObjects)).toBe(2)
})
