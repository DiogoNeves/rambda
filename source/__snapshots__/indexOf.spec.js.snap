// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brute force Array | Boolean 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | Boolean 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [],
    ],
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | Boolean 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
    ],
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | Null 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Null 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [],
    ],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Null 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
    ],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Object 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | Object 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [],
    ],
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | Object 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
    ],
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | String 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 0,
  "result": "PENDING",
}
`;

exports[`brute force Array | String 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [],
    ],
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 0,
  "result": "PENDING",
}
`;

exports[`brute force Array | String 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
    ],
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Array | Undefined 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Undefined 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [],
    ],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Undefined 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      Array [
        1,
      ],
      Array [
        2,
      ],
    ],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Boolean 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Function | Null 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Object 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Function | String 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Function | Undefined 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Object | Boolean 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | Boolean 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | Boolean 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 2,
    },
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | Null 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Object | Null 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Object | Null 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 2,
    },
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Object | Object 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | Object 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | Object 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 2,
    },
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | String 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | String 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | String 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 2,
    },
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Object | Undefined 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Object | Undefined 2`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 1,
    },
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Object | Undefined 3`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Object {
      "a": 2,
    },
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Promise | Boolean 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    Promise {},
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Promise | Null 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    Promise {},
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Promise | Object 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    Promise {},
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Promise | String 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    Promise {},
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force Promise | Undefined 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    Promise {},
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force RegExp | Boolean 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    /foo/,
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force RegExp | Null 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    /foo/,
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force RegExp | Object 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    /foo/,
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force RegExp | String 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    /foo/,
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force RegExp | Undefined 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    /foo/,
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String | Boolean 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of true",
    "type": "Error",
  },
  "inputs": Array [
    "foo",
    true,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force String | Null 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "Error",
  },
  "inputs": Array [
    "foo",
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of null",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String | Object 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of [object Object]",
    "type": "Error",
  },
  "inputs": Array [
    "foo",
    Object {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": -1,
  "result": "PENDING",
}
`;

exports[`brute force String | String 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of foo",
    "type": "Error",
  },
  "inputs": Array [
    "foo",
    "foo",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 0,
  "result": "PENDING",
}
`;

exports[`brute force String | Undefined 1`] = `
Object {
  "error": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "Error",
  },
  "inputs": Array [
    "foo",
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": Object {
    "message": "Cannot read property 'indexOf' of undefined",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;
